ui/src/image_annotator.tsx

```diff
@@ -248,56 +388,75 @@ export const XImageAnnotator = ({ model }: { model: ImageAnnotator }) => {
-          break
-        }
-        case 'polygon': {
-          redrawExistingShapes()
-          polygonRef.current?.drawPreviewLine(cursor_x, cursor_y, getCurrentTagColor(activeTag))
-          if (polygonRef.current?.isIntersectingFirstPoint(cursor_x, cursor_y)) canvas.style.cursor = 'pointer'
-          break
+        { cursor_x, cursor_y } = eventToCursor(e, canvas.getBoundingClientRect(), zoom, imgPositionRef.current),
+        clickStartPosition = clickStartPositionRef.current
+
+      if (e.ctrlKey && zoom > 1) {
+        canvas.style.cursor = clickStartPosition?.dragging ? 'grabbing' : 'grab'
+        if (clickStartPosition?.dragging && imgCanvasRef.current) {
+          translateImage((cursor_x - clickStartPosition.x) * zoom, (cursor_y - clickStartPosition.y) * zoom, zoom)
+          updateCanvas(zoom)
         }
-        case 'select': {
-          // If left mouse btn is not held during moving, ignore.
-          if (e.buttons !== 1) break
-          if (focused?.shape.rect) {
-            rectRef.current?.onMouseMove(cursor_x, cursor_y, focused, intersected, clickStartPosition)
+      } else {
+        const
+          focused = drawnShapes.find(({ isFocused }) => isFocused),
+          intersected = getIntersectedShape(drawnShapes, cursor_x, cursor_y)
+        canvas.style.cursor = getCorrectCursor(cursor_x, cursor_y, focused, intersected, activeShape === 'select')
+        switch (activeShape) {
+          case 'rect': {
+            const currentlyDrawnRect = rectRef.current?.onMouseMove(cursor_x, cursor_y, focused, intersected, clickStartPosition)
             redrawExistingShapes()
+            if (currentlyDrawnRect?.rect) rectRef.current?.drawRect(currentlyDrawnRect.rect, getCurrentTagColor(activeTag))
+            break
           }
-          else if (focused?.shape.polygon) {
-            polygonRef.current?.onMouseMove(cursor_x, cursor_y, focused, intersected, clickStartPosition)
+          case 'polygon': {
             redrawExistingShapes()
+            polygonRef.current?.drawPreviewLine(cursor_x, cursor_y, getCurrentTagColor(activeTag))
+            if (polygonRef.current?.isIntersectingFirstPoint(cursor_x, cursor_y)) canvas.style.cursor = 'pointer'
+            break
+          }
+          case 'select': {
+            // If left mouse btn is not held during moving, ignore.
+            if (e.buttons !== 1) break
+            const shape = focused?.shape.rect ? rectRef.current : polygonRef.current
+            // TODO: Move all selected shapes at once.
+            if (shape) {
+              shape.onMouseMove(cursor_x, cursor_y, focused, intersected, clickStartPosition)
+              // Deselect all other shapes when one starts moving.
+              if (drawnShapes.some(ds => ds.isFocused && ds !== focused)) {
+                setDrawnShapes(drawnShapes => drawnShapes.map(ds => ({ ...ds, isFocused: intersected === ds })))
+              }
```

Comment:
No need to go through the drawnShapes twice, we can update anyway.

```suggestion
                setDrawnShapes(drawnShapes => drawnShapes.map(ds => ({ ...ds, isFocused: intersected === ds })))```